bin/get-video-s3-acls.py

```diff
@@ -0,0 +1,409 @@
+):
+
+    action = ""
+    # Cases
+    # In S3     In NZSL     Action
+    #   Is        No          Delete!
+    #   Is        Is          Check ACL
+    #   Not       Is          Review
+    #   (F F impossible)
+
+    if key_in_s3:
+        if key_in_nzsl:
+            action = "Check ACL"
+        else:
+            action = "Delete"
+    else:
+        action = "Review"
+
+    # See signbank/video/models.py, line 59, function set_public_acl()
+    if key_in_nzsl:
+        canned_acl_expected = "public-read" if video_public else "private"
+    else:
+        canned_acl_expected = ""
+
+    # If key not in S3, just return its NZSL info and action
+    if not key_in_s3:
+        return CSV_DELIMITER.join(
+            [
+                f"{video_key}",
+                f"{gloss_idgloss}",
+                f"{gloss_created_at}",
+                "",  # S3 LastModified
+                f"{canned_acl_expected}",
+                "",  # Actual Canned ACL
+                f"{gloss_id}",
+                f"{video_id}",
+                f"{gloss_public}",
+                f"{video_public}",
+                action,
+            ]
+        )
+
+    # Get S3 object's ACL
+    result = subprocess.run(
+        [
+            AWSCLI,
+            "s3api",
+            "get-object-acl",
+            "--output",
+            "json",
+            "--bucket",
+            AWS_S3_BUCKET,
+            "--key",
+            video_key,
+        ],
+        env=os.environ,
+        shell=False,
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    canned_acl = "unknown"
+    acls_grants_json = json.loads(result.stdout)["Grants"]
+    if len(acls_grants_json) > 1:
+        if (
+            acls_grants_json[0]["Permission"] == "FULL_CONTROL"
+            and acls_grants_json[1]["Permission"] == "READ"
+        ):
+            canned_acl = "public-read"
+    elif acls_grants_json[0]["Permission"] == "FULL_CONTROL":
+        canned_acl = "private"
+
+    # Get S3 object's LastModified date/time
+    result = subprocess.run(
+        [
+            AWSCLI,
+            "s3api",
+            "get-object-attributes",
+            "--object-attributes",
+            "ObjectParts",
+            "--output",
+            "json",
+            "--bucket",
+            AWS_S3_BUCKET,
+            "--key",
+            video_key,
+        ],
+        env=os.environ,
+        shell=False,
+        check=True,
+        capture_output=True,
+        text=True,
+    )
+    s3_lastmodified = json.loads(result.stdout)["LastModified"]
+
+    return CSV_DELIMITER.join(
+        [
+            f"{video_key}",
+            f"{gloss_idgloss}",
+            f"{gloss_created_at}",
+            f"{s3_lastmodified}",
+            f"{canned_acl_expected}",
+            f"{canned_acl}",
+            f"{gloss_id}",
+            f"{video_id}",
+            f"{gloss_public}",
+            f"{video_public}",
+            action,
+        ]
+    )
```

Comment:
suggestion: encapsulating these into functions should greatly simplify this here

The responsibility of the `build_csv_row` function is to build a valid CSV row from the given data - it primarily cares about including "some data" in "some order", but _not_ what the actual data is (aside from that it's valid in a CSV context); i.e. it doesn't need to know that `s3_lastmodified` value comes from a `LastModified` property from a subprocess call, or the logic behind how `canned_acl` is determined

(generally this is acceptable if the logic is small, which is why e.g. it's fine to keep the `key_in_nzsl` logic in this function)

Reply:
I refactored into functions but it might be nice to completely decouple things, maybe via lists. Depends how tricky we want to get.  I'll mark this one resolved because the code has been rearranged.  If you want further refactoring just make new change requests.